add test cases, to do match up test between sentinel and helios.

yinghuang123 opened the pull request:
Required items, please complete
   
   Change(s):
   - See guidance below

   Reason for Change(s):
   - See guidance below

   Version Updated:
   - Required only for Detections/Analytic Rule templates
   - See guidance below

   Testing Completed:
   - See guidance below

   Checked that the validations are passing and have addressed any issues that are present:
   - See guidance below


# Guidance <- remove section before submitting
-----------------------------------------------------------------------------------------------------------
## Before submitting this PR please ensure that you have read the following sections and filled out the changes, reason for change and testing complete sections:

Thank you for your contribution to the Microsoft Sentinel Github repo.

> Details of the code changes in your submitted PR.  Providing descriptions for pull requests ensures there is context to changes being made and greatly enhances the code review process.  Providing associated Issues that this resolves also easily connects the reason.
   
   Change(s):
   - Updated syntax for XYZ.yaml

   Reason for Change(s):
   - New schema used for XYZ.yaml
   - Resolves ISSUE #1234

   Version updated:
   - Yes
   - Detections/Analytic Rule templates are required to have the version updated

> The code should have been tested in a Microsoft Sentinel environment that does not have any custom parsers, functions or tables, so that you validate no incorrect syntax and execution functions properly.  If your submission requires a custom parser or function, it must be submitted with the PR.

   Testing Completed:
   - Yes/No/Need Help

_Note: If updating a detection, you must update the version field._

> Before the submission has been made, please look at running the KQL and Yaml Validation Checks locally.
> https://github.com/Azure/Azure-Sentinel#run-kql-validation-locally

   Checked that the validations are passing and have addressed any issues that are present:
   - Yes/No/Need Help
   
   _Note: Let us know if you have tried fixing the validation error and need help._

> References: 
> - [Guidance for Detection checks](https://github.com/Azure/Azure-Sentinel#pull-request-detection-template-structure-validation-check)
> - [General contribution guidance](https://github.com/Azure/Azure-Sentinel/wiki#what-can-you-contribute-and-how-can-you-create-contributions)
> - [PR validation troubleshooting](https://github.com/Azure/Azure-Sentinel#pull-request)


-----------------------------------------------------------------------------------------------------------

yinghuang123 committed:
add test cases, to do match up test between sentinel and helios.

eerus commented on Solutions/CohesitySecurity/Tests/az.test.py:
```diff
@@ -1,16 +1,65 @@
+
+    alert_details = get_alert_details(alert_id, apiKey)
+    print("alert_id --> %s" % alert_id)
+    print("apiKey --> %s" % apiKey)
```
Do we really need to publish the API key (data that can be used for intrusion)? If we do, then maybe you can limit it to first 4 and last 4 symbols?

yinghuang123 replied:
this is just local printing, which should be fine.
the intruder could still get it from local cohesity.json file.
we need this info to debug, when test case failed.

eerus reviewed (changes requested):
LGTM overall but I left one security comment.

kishannerella commented on Solutions/CohesitySecurity/Playbooks/Cohesity_Close_Helios_Incident/azuredeploy.json.sh:
```diff
@@ -0,0 +1,12 @@
+#!/bin/zsh
```
Can we move this files into Tools/ directory instead? This should be generic script which take the name of the playbook as input and deploys the playbook. I don't think we need one such file for every playbook. Also, as much as possible let's try to seperate development folders and test/script folders.

Please add a description on the top of every new file that we add. We should know what the file is supposed to do by looking at the header.

yinghuang123 replied:
just check, till now, for all of the playbooks we have, they just need one common/same parameter, that's PlaybookName.
so, we could use one sh to serve all.

I will just "hide" this file from common/public-visible branch, for my vrun function of vim, will look for a nearby "azuredeploy.json.sh" file to run, when I press F5 key in azuredeploy.json file.

kishannerella commented on Solutions/CohesitySecurity/Tests/az.py:
```diff
@@ -42,3 +42,22 @@ def has_dup_incidents(resource_group, workspace_name):
     return [(jsObj["id"].split("/")[-1], jsObj["description"].split("Helios ID: ")[-1]) for jsObj in json.loads(result.stdout)]
+
+
+def search_alert_id_in_incident(alert_id, resource_group, workspace_name):
```
Please add a comment to this function and other functions in this file clearly mentioning what they're supposed to do and what the return value is supposed to be. I know that in some cases it's trivial, but let's have some conventions which will help in readability and maintainability.

kishannerella replied:
Also, please add a header in the file describing what this file is supposed to do.

yinghuang123 replied:
done

kishannerella commented on Solutions/CohesitySecurity/Tests/helios.py:
```diff
@@ -1,12 +1,47 @@
+    return response.json() if response.json() else None
 
 
 def get_alert_details(alert_id, apiKey):
```
Please remove this function and let the users always use always use the newly created function. We can remove the duplicated code that way.

yinghuang123 replied:
done

kishannerella commented on Solutions/CohesitySecurity/Tests/helios.py:
```diff
@@ -1,12 +1,47 @@
+
+
+def get_alerts_details(alert_ids, apiKey):
+    maxAlerts = len(alert_ids)
```
Assert that len() is > 0

yinghuang123 replied:
done

kishannerella commented on Solutions/CohesitySecurity/Tests/helios.py:
```diff
@@ -1,12 +1,47 @@
+    return response.json()[0] if response.json() else None
+
+
+def get_alerts(apiKey, startDaysAgos=30, endDaysAgos=10):
```
Please rename the variables
startDaysAgos -> startDaysAgo
endDaysAgos -> endDaysAgo

Let the default value for endDaysAgo be '0' as that's what we use.

yinghuang123 replied:
changed to start_days_ago and end_days_ago.

kishannerella commented on Solutions/CohesitySecurity/Tests/helios.py:
```diff
@@ -1,12 +1,47 @@
+        return str(days_ago_timestamp)
+
+    api_url = "https://helios.cohesity.com/mcm/alerts?alertCategoryList=kSecurity&startDateUsecs=" + get_days_ago_timestamp(startDaysAgos) + "&endDateUsecs=" + get_days_ago_timestamp(endDaysAgos)
+    headers = {"Content-Type": "application/json"}
```
Please create a function for 'headers' construction. This is duplicate code between get_alerts_details and get_alerts

yinghuang123 replied:
done

kishannerella commented on Solutions/CohesitySecurity/Tests/helios.py:
```diff
@@ -1,12 +1,47 @@
+
+
+def get_alerts(apiKey, startDaysAgos=30, endDaysAgos=10):
+    def get_days_ago_timestamp(daysAgos=10):
```
Consider removing the default value to avoid mistakes in calling this function. We should only use this for a value that we know.

kishannerella replied:
This is not yet resolved.

yinghuang123 replied:
done

kishannerella commented on Solutions/CohesitySecurity/Tests/az.py:
```diff
@@ -42,3 +42,22 @@ def has_dup_incidents(resource_group, workspace_name):
+
+
+def search_alert_id_in_incident(alert_id, resource_group, workspace_name):
+    response = subprocess.run(['az', 'sentinel', 'incident', 'list',
```
Question: When this script is run, are the logs stored somewhere so that it's useful for debugging? Otherwise, it's difficult to debug.

kishannerella replied:
Could you please answer this question?

yinghuang123 replied:
this is a one time test code, not daemon or long time running server application.
no need to Permanent the log.
just write to command line output is enough.

kishannerella replied:
As long the response is printed, I'm good.

kishannerella commented on Solutions/CohesitySecurity/Tests/az.test.py:
```diff
@@ -1,16 +1,65 @@
 apiKey = data['apiKey']
 f.close()
 
 assert has_dup_incidents(resource_group, workspace_name) == False
```
Please move this [down](https://github.com/cohesity/Azure-Sentinel/pull/8/commits/0181cc6743413b2fc1a3ef975b8b38ed4a8f0331#diff-962835dddd48b7def1a7bed8c5b7a8ec6391a111f42c8f1da2cc8f13bbc70cb7R62),  so that  it's clubbed with the testing functions. Infact, please move all the above code to the end of the file so that there's clear seperation between functions and when we're calling.

yinghuang123 replied:
done

kishannerella commented on Solutions/CohesitySecurity/Tests/az.test.py:
```diff
@@ -1,16 +1,65 @@
 resource_group = data['resource_group']
 workspace_name = data['workspace_name']
-playbook_name = data['playbook_name']
 apiKey = data['apiKey']
```
Could we please validate that none of these values are empty so that we fail fast and the user knows that their environment is not configured?

yinghuang123 replied:
we don't need to validate/test against test codes, this is over test.
the user just need to fill the cohesity.json file, which is also used during cli deployment.           
it is very rarely that the user would leave some attributes empty.

kishannerella replied:
This kind of defensive programming is used all across Cohesity. Could you tell me if not having 'apiKey' in the json will exit here immediately? Or will it proceed further with empty value? If it proceeds without any empty value, then I would insist on assertion. If it fails immediately, then I'll leave it up to you whether to add assert on empty or not.

yinghuang123 replied:
done.

yinghuang123 replied:
could you see my comments above?

kishannerella commented on Solutions/CohesitySecurity/Tests/helios.py:
```diff
@@ -1,12 +1,47 @@
 
 import requests
+import datetime
+import datetime
```
Remove duplicate?

yinghuang123 replied:
done

kishannerella commented on Solutions/CohesitySecurity/Tests/az.test.py:
```diff
@@ -1,16 +1,65 @@
+
+def test_cohesity_close_helios_incident():
+    playbook_name = "Cohesity_Close_Helios_Incident"
+    subscriptionId = get_subscriptionId()
```
s/subscriptionId/subscription_id

yinghuang123 replied:
done, will be visible after next push.

kishannerella commented on Solutions/CohesitySecurity/Tests/helios.py:
```diff
@@ -1,12 +1,47 @@
+import time
+
+
+def get_alerts_details(alert_ids, apiKey):
```
We should not be needing to pass around apiKey for every function. Ideally, we should have the functions in this file clubbed into a class which accepts the API key in the constructor and that should be re-used. I'll leave it upto you if you want to do it now or later.

yinghuang123 replied:
yes, we could have a class which provide get_*() functions to get all the required field.
while this is just some test codes. 
I don't feel very strong to do this abstraction right now.

kishannerella commented on Solutions/CohesitySecurity/Tests/az.test.py:
```diff
@@ -1,16 +1,65 @@
 resource_group = data['resource_group']
 workspace_name = data['workspace_name']
-playbook_name = data['playbook_name']
 apiKey = data['apiKey']
```
s/apiKey/api_key

Please remove all such references in other files to be consistent.

yinghuang123 replied:
done

yinghuang123 committed:
Merge remote-tracking branch 'cohesity/CohesitySecurity.internal' into automation_test

yinghuang123 committed:
hide/remove this file

yinghuang123 committed:
some small changes according to review feedback.

kishannerella commented on Solutions/CohesitySecurity/Tests/az.py:
```diff
@@ -1,4 +1,7 @@
 #!/usr/bin/env python3
+'''
+provide wrapper functions for azure cli to interact with azure sentinel.
```
Please replace with

"This file contains the wrapper functions to invoke Azure cli to interact with Azure Sentinel"

yinghuang123 replied:
I changed it to "This module provides wrapper functions for Azure CLI to interact with Azure Sentinel."
how about this?

kishannerella commented on Solutions/CohesitySecurity/Tests/az.py:
```diff
@@ -8,23 +11,37 @@
-def run_playbook(subscriptionId, vid, resource_group, workspace_name, playbook_name):
-    result = subprocess.run(['az', 'sentinel', 'incident', 'run-playbook', '--incident-identifier', vid, '--resource-group', resource_group, '--workspace-name', workspace_name, '--logic-apps-resource-id', "/subscriptions/" + subscriptionId + "/resourceGroups/" + resource_group + "/providers/Microsoft.Logic/workflows/" + playbook_name], stdout=subprocess.PIPE)
+'''
+run a playbook by name, resource_group and workspace_name
```
nit: Please start sentences with a capital letter.

kishannerella replied:
Same for all the lines in the function and other functions.

yinghuang123 replied:
done

kishannerella commented on Solutions/CohesitySecurity/Tests/az.test.py:
```diff
@@ -8,44 +8,37 @@
 import os
 
 os.chdir(os.path.dirname(os.path.abspath(__file__)))
```
Please move this line towards the end as well.

yinghuang123 replied:
no, it has to be at the beginning.
for I used relative path in code, if move to the end, all those test cases would fail.

kishannerella replied:
1. I'm sorry. I should have been more clear. I didn't mean at the end of the file. You moved all the remaining code to the end of the file based on my comment. Why is this the only line that's not moved?
2. We should not use chdir in code. It's a bad practice. All paths should be based on root dir which should be based on an environment variable, something like COHESITY_SECURITY_ROOT_DIR or an argument to the function.

yinghuang123 replied:
I respectfully disagree with the idea of defining a root directory for the project. While it may be useful in some cases, it could potentially introduce dependencies and make the code less portable. Additionally, if all paths are based on the root directory, it may limit the script's flexibility and require the script to be run from a specific location. Moreover, adding another environment variable could make the project unnecessarily complicated.

kishannerella replied:
Entire Cohesity code uses this format and yes, things are always run from the root directory. In this case, technically you don't need to do chdir. All you have to do is store the current directory in a variable and use the variable to open files using relative file paths. But it's fine if you want to do it this way.

kishannerella commented on Solutions/CohesitySecurity/Tests/az.py:
```diff
@@ -33,17 +50,26 @@ def get_one_incident_id(resource_group, workspace_name):
+'''
+to check whether there is any dup incident in the specified resource_group and workspace_name.
+'''
 def has_dup_incidents(resource_group, workspace_name):
```
This function should be moved to az.test.py

kishannerella commented on Solutions/CohesitySecurity/Tests/az.test.py:
```diff
@@ -59,6 +52,14 @@ def test_alerts_in_sentinel():
+data = json.load(f)
+resource_group = data['resource_group']
+workspace_name = data['workspace_name']
+api_key = data['apiKey']
```
Why is apiKey in camel case but not resource_group or workspace_name? Please try to be consistent.

yinghuang123 replied:
apiKey is in data, not variable name. it should be fine.
impossible to be 100% consistent, for apiKey is also used in helios endpoint request header...
    -H 'apiKey: 33e44eac-ce99-46df-7f4e-9ac39446a66e'
could we change that? 

of course I also want to be 100% consistent..
that's why I used apiKey in python code, to be consistent with helios endpoint request header

kishannerella reviewed (commented):
Mostly LG. Added some minor comments. Will approve once those are resolved.

yinghuang123 committed:
add this file, for further customization.
+++ ./Solutions/CohesitySecurity/Tools/createSolutionV2.ps1

yinghuang123 committed:
add these files to parse alert.json
+++ ./Solutions/CohesitySecurity/Tests/Alert.py
+++ ./Solutions/CohesitySecurity/Tests/Alert.test.py
some other small changes.

yinghuang123 committed:
Merge remote-tracking branch 'cohesity/automation_test' into automation_test

yinghuang123 committed:
add/update comments in codes.
change from apiKey to api_key

yinghuang123 committed:
small changes based on review feedback.
add/update comments.

kishannerella commented on Solutions/CohesitySecurity/Tests/Alert.py:
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env python3
```
Please add a description for the file.

yinghuang123 replied:
done

kishannerella commented on Solutions/CohesitySecurity/Tests/Alert.py:
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env python3
+import json
+
```
Is there a reason to name the file starting with a capital letter, not in sync with other files in the same directory?

yinghuang123 replied:
done

kishannerella commented on Solutions/CohesitySecurity/Tests/Alert.py:
```diff
@@ -0,0 +1,28 @@
+                self.clusterIncarnationId = prop['value']
+        self.protectionGroupId = f"{self.clusterId}:{self.clusterIncarnationId}:{self.jobId}"
+
+    def get_jobId(self):
```
Why is the name 'get_jobId' a mix of snake case and camel case? Same for the other function names as well.

yinghuang123 replied:
done

kishannerella commented on Solutions/CohesitySecurity/Tests/Alert.py:
```diff
@@ -0,0 +1,28 @@
+                self.jobId = prop['value']
+            elif prop['key'] == 'clusterIncarnationId':
+                self.clusterIncarnationId = prop['value']
+        self.protectionGroupId = f"{self.clusterId}:{self.clusterIncarnationId}:{self.jobId}"
```
Please assert if jobId or clusterIncarnationId or clusterId is empty here after parsing from the data.

I want to understand what happens in the case where the propertyList doesn't have jobId or clusterIncarnationId? Will there be a run time fault? Or will be default initialized to empty string?

yinghuang123 replied:
done

kishannerella commented on Solutions/CohesitySecurity/Tests/Alert.test.py:
```diff
@@ -0,0 +1,121 @@
+#!/usr/bin/env python3
```
Please add a description to the file.

Also, why does the name of the file start with a capital letter?

yinghuang123 replied:
done

kishannerella commented on Solutions/CohesitySecurity/Tests/Alert.test.py:
```diff
@@ -0,0 +1,121 @@
+
+        self.alert = Alert(self.json_str)
+
+    def test_get_jobId(self):
```
Same issue as before. Mix of snake case and camel case.

kishannerella commented on Solutions/CohesitySecurity/Tests/az.py:
```diff
@@ -1,76 +1,64 @@
+    Returns the ID and alert ID of a random incident from the list of incidents.
+    """
     result = subprocess.run(['az', 'sentinel', 'incident', 'list', '--resource-group', resource_group, '--workspace-name', workspace_name], stdout=subprocess.PIPE)
     jsObj = random.choice(json.loads(result.stdout))
```
What happens in case there are 0 incidents? Will this result in a run time fault? Can we add an assert?

kishannerella commented on Solutions/CohesitySecurity/Tests/az.test.py:
```diff
@@ -1,66 +1,102 @@
+class TestCohesity(unittest.TestCase):
+    def setUp(self):
+        """
+        This method is called before each test case and is used to set up any necessary resources or configurations.
```
These lines are big. Could you please split it? Generally, we try to have a 80-char limit for coding. Is there a limit that we already imposed? If not, let's use the 79 char limit from now on. It's used within Cohesity in C++ and Go.

kishannerella replied:
Missed this?

yinghuang123 replied:
done

kishannerella commented on Solutions/CohesitySecurity/Tests/az.test.py:
```diff
@@ -1,66 +1,102 @@
+
+    def test_cohesity_close_helios_incident(self):
+        """
+        This test case verifies that the 'Cohesity_Close_Helios_Incident' playbook can successfully close a Helios
```
Same comment as before. Line is too long.

kishannerella replied:
Missed this?

yinghuang123 replied:
done

kishannerella commented on Solutions/CohesitySecurity/Tests/az.test.py:
```diff
@@ -1,66 +1,102 @@
+
+
+class TestCohesity(unittest.TestCase):
+    def setUp(self):
```
s/setUp/setup/

yinghuang123 replied:
after making this suggested change, this function would not be executed......
seems this is syntax restriction...
undo my change.

kishannerella replied:
Ack

kishannerella commented on Solutions/CohesitySecurity/json_parser.sh:
```diff
@@ -6,7 +6,7 @@ cd "$SCRIPTPATH"
 ClientId=$(cat ./cohesity.json | jq '."ClientId"' | sed 's/^"//g;s/"$//g')
```
In the next PR, please update the other variables in cohesity.json and json_parser.sh to follow the snake case like 'api_key'.

kishannerella commented on Solutions/CohesitySecurity/Tests/az.test.py:
```diff
@@ -1,66 +1,102 @@
+        # self.skipTest("Skipping test_cohesity_close_helios_incident")
+        playbook_name = "Cohesity_Close_Helios_Incident"
+        subscription_id = get_subscription_id()
+        vid, alert_id = get_one_incident_id(self.resource_group, self.workspace_name)
```
Could you please use a more expressive name than vid? I'm not sure what vid means here.

yinghuang123 replied:
done

kishannerella commented on Solutions/CohesitySecurity/Tests/az.test.py:
```diff
@@ -1,66 +1,102 @@
+        return len(alert_ids) != len(np.unique(np.array(alert_ids)))
+
+    def test_alerts_in_sentinel(self):
+        """
```
Can you please add log messages for each test mentioning that the test is starting and finished successfully so that someone looking at the output can understand the current status of the test?

Also, in general, since this is testing, it's a good idea to print the output of the commands or the results we get from API responses to aid debugging. When you do this, please ensure that the timestamps are also present in the logs. You can do the latter part in a seperate PR if you wish, but please ensure the first paragraph of my comment is implemented in this PR>

yinghuang123 replied:
the first part is done.

yinghuang123 replied:
also applied this change to another test.py

yinghuang123 committed:
rename files.
add comments, small changes according to review feedback.

yinghuang123 committed:
add this renamed file.

yinghuang123 committed:
remove old file

yinghuang123 committed:
add the new renamed file.

yinghuang123 committed:
small changes.

kishannerella commented on Solutions/CohesitySecurity/Tests/az.test.py:
```diff
@@ -67,34 +68,44 @@ def test_cohesity_close_helios_incident(self):
+        print("test_cohesity_close_helios_incident finished successfully.")
+
+
+def test_all_incidents_in_helios(self):
```
Why is the definition test_all_incidents_in_helios and the functions below are not within the scope of the class TestCohesity?

yinghuang123 replied:
done

yinghuang123 committed:
fix the indent issue.

yinghuang123 committed:
small changes based on feedback.
